More unity

permcody opened the pull request:
By default lump all source at each top level directory into single unity files. This is a departure from before when we did one unity file per directory.

closes #26028

permcody committed:
More unity

By default lump all source at each top level directory into single unity files.
This is a departure from before when we did one unity file per directory.

closes #26028

dschwen commented:
This is not going to work. We make use of subdirectories deliberately to avoid lumping stuff in with the rest of the unity sources for libtorch for example.

permcody commented:
The capability to remove individual directories from being “unitied” still exists. Can we not just add those directories explicitly to the list where needed?

YaqiWang commented on framework/app.mk:
```diff
@@ -79,14 +79,16 @@ $(eval $(call unity_dir_rule, $(unity_src_dir)))
+# of unrelated stuff.
+# Note: If a top-level directory is excluded from being built with unity, we
+# ignore contents of all subdirectories under that folder.
+non_unity_dirs := %.libs %/src $(app_non_unity_dirs)
```
How about utilize a new variable like `app_level1_unity_dirs`? It will be more complicated than this simple change, but it can reserve the current behavior.

permcody replied:
There’s already an application level variable you can set.

YaqiWang replied:
Are you saying that non-unity variable? This new variable is to specify what directories should be considered in unity build but lump all files under it including its sub-folders. Maybe I misunderstood something?

YaqiWang replied:
I was suggesting something like this:
```
diff --git a/framework/app.mk b/framework/app.mk
index 4c7672576c..6251e6f014 100644
--- a/framework/app.mk
+++ b/framework/app.mk
@@ -80,7 +80,7 @@ $(eval $(call unity_dir_rule, $(unity_src_dir)))
 # that won't benefit from unity building
 # Also, exclude the base directory by default because it's another big jumble
 # of unrelated stuff
-non_unity_dirs := %.libs %/src %src/base $(app_non_unity_dirs)
+non_unity_dirs := %.libs %/src $(app_non_unity_dirs)
 
 # Find all of the individual subdirectories
 # We will create a Unity file for each individual subdirectory
@@ -88,8 +88,11 @@ non_unity_dirs := %.libs %/src %src/base $(app_non_unity_dirs)
 # and will benefit from a Unity build
 srcsubdirs := $(shell find $(APPLICATION_DIR)/src -type d -not -path '*/.libs*')
 
+level1_srcsubdirs := $(shell find $(app_level1_unity_dirs) -type d -not -path '*/.libs*')
+
 # Filter out the paths we don't want to Unity build
 unity_srcsubdirs := $(filter-out $(non_unity_dirs), $(srcsubdirs))
+unity_srcsubdirs := $(filter-out $(level1_srcsubdirs), $(unity_srcsubdirs))
 non_unity_srcsubdirs := $(filter $(non_unity_dirs), $(srcsubdirs))
 
 # This is a biggie
@@ -98,8 +101,11 @@ non_unity_srcsubdirs := $(filter $(non_unity_dirs), $(srcsubdirs))
 # full hierarchy of the path underneath src
 $(foreach srcsubdir,$(unity_srcsubdirs),$(eval $(call unity_file_rule,$(call unity_unique_name,$(unity_src_dir),$(APPLICATION_DIR),$(srcsubdir)),$(shell find $(srcsubdir) -maxdepth 1 \( -type f -o -type l \) -regex "[^\#~]*\.C"),$(srcsubdir),$(unity_src_dir))))
 
+$(foreach srcsubdir,$(app_level1_unity_dirs),$(eval $(call unity_file_rule,$(call unity_unique_name,$(unity_src_dir),$(APPLICATION_DIR),$(srcsubdir)),$(shell find $(srcsubdir) \( -type f -o -type l \) -regex "[^\#~]*\.C"),$(srcsubdir),$(unity_src_dir))))
+
 # This creates the whole list of Unity source files so we can use it as a dependency
 app_unity_srcfiles := $(foreach srcsubdir,$(unity_srcsubdirs),$(call unity_unique_name,$(unity_src_dir),$(APPLICATION_DIR),$(srcsubdir)))
+app_unity_srcfiles += $(foreach srcsubdir,$(app_level1_unity_dirs),$(call unity_unique_name,$(unity_src_dir),$(APPLICATION_DIR),$(srcsubdir)))
 
 # Add to the global list of unity source files
 unity_srcfiles += $(app_unity_srcfiles)
```

permcody replied:
@YaqiWang - This PR already changes the existing behavior so that we lump all depth = 1 data together across the board, which is exactly what you want for Griffin. It halves the compile time. It's most of the way there! I think the only thing we need to address now are exceptions. For that we have `$app_non_unity_dirs` which should do the right thing if we need to exclude a specific directory in an application. 

IMO - We don't need any other logic at the top level, as @dschwen pointed out, I'll need to exclude the libTorch directories, but hopefully that's about it unless there are other application level conflicts.

Am I going to finish this up? Probably won't get another chance until about the middle of the December! If nobody else gets to it before then, I'll pick it back up!

permcody replied:
@YaqiWang - This PR already changes the existing behavior so that we lump all depth = 1 data together across the board, which is exactly what you want for Griffin. It halves the compile time. It's most of the way there! I think the only thing we need to address now are exceptions. For that we have $app_non_unity_dirs which should do the right thing if we need to exclude a specific directory in an application.

IMO - We don't need any other logic at the top level, as @dschwen pointed out, I'll need to exclude the libTorch directories, but hopefully that's about it unless there are other application level conflicts.

Am I going to finish this up? Probably won't get another chance until about the middle of the December! If nobody else gets to it before then, I'll pick it back up!

YaqiWang commented:
@permcody any plan on finishing this?

permcody commented:
Timings

Before: `make -j10 1286.05s user 196.03s system 569% cpu 4:20.24 total`
After: `make -j10 691.59s user 117.54s system 464% cpu 2:54:05 total`

permcody committed:
More tweaks to unity logic

Here we want to ensure that we pickup source files of subdirectories of non_unity directories.
To do that, we'll run a separate find to get all subdirectories of a location to filter against.

refs #26028

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1958274/) on 3d07993 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/26051/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1958289/) on 3d07993 wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/26051/coverage/xfem)

</details>



This comment will be updated on new commits.

moosebuild commented:
Job [Conda (Ubuntu)](https://civet.inl.gov/job/1977007/) on 28971df : invalidated by @permcody

Can&#x27;&#x27;t replicate

moosebuild commented:
Job [Documentation internal apps](https://civet.inl.gov/job/1977000/) on 28971df : invalidated by @permcody

moosebuild commented:
Job [Documentation public apps](https://civet.inl.gov/job/1977001/) on 28971df : invalidated by @permcody

moosebuild commented:
Job [Public app tests](https://civet.inl.gov/job/1977003/) on 28971df : invalidated by @permcody

moosebuild commented:
Job [Conda MOOSE (Linux)](https://civet.inl.gov/job/1976979/) on 28971df : invalidated by @permcody

moosebuild commented:
Job [OpenMPI](https://civet.inl.gov/job/1977010/) on 28971df : invalidated by @permcody

permcody commented:
I've opened issue https://github.com/idaholab/moose/issues/26508 to look at the revision generation issue separately. It is not related to this PR.

permcody commented:
> This is not going to work.

Narrator: ... but it did work, and they all lived happily ever after!
